### ISSUE TITLE
REST API Issues

### ISSUE THREAD

**Issue by donnapep** [Automattic/wp-super-cache#229]
The following settings all return `null` instead of a value:
`cache_mobile_browsers` - lists the mobile browsers when "Mobile device support" is enabled.
`cache_mobile_prefixes` - lists the mobile prefixes when "Mobile device support" is enabled.
`cache_mutex_disabled` - "Coarse file locking" setting.
`cache_hello_world` - "Proudly tell the world your server is Stephen Fry proof" setting.
`cache_gzencode` - Shows a warning that compression is disabled if gzencode function was not found.
`cache_disable_locking` - Show / hide "Coarse file locking" setting.
`cache_readonly` - Shows a warning that the cache path is not read-only.
`cache_writable` - Shows a warning that the cache path is currently writable.

The following settings are missing from the API:
`cache_gc_email_me` - "Email me when the garbage collection runs" setting.
`cache_schedule_interval` - Value of "Interval" setting when clock scheduling is enabled.
`cache_next_gc` - Value of "Next scheduled garbage collection".
`accepted_files` - Accepted filenames.
`rejected_uri` - Rejected URIs.
`rejected_user_agent` - Rejected user agents.
`lock_down` - Flag that determines what text is displayed in the "Lock Down" section.
`cache_direct_pages` - Directly cached files.
`http_only` - This setting is used for the cache tester functionality on the Easy tab and should return a default value of `true`.

_Cache Stats_
For caching stats, `cached_list` and `expired_list` for both `supercache` and `wpcache` are objects where age is the key. Could we return these as an array of objects instead where each object has the following structure:

`{ age: 1000, uri: "donnapep.wpsandbox.me/blog/blah" }`

_Cache Type_
`is_super_cache_enabled` is always returned as `false` by the API regardless of what caching type is selected. The strange thing is that the test harness *does* return the correct value, but the network request in Calypso does not.

_Saving_
POSTing data does not work at all.

**Comment by dbtlr** [comment]
I've added missing settings in #215, along with the following:

- `cache_disable_locking` AND `cache_mutex_disabled` are the same thing, I've merged them into just `cache_disable_locking`.
- `cache_readonly` AND `cache_writable` are notices returned by the notices endpoint.
- `httponly` doesn't have a persistence setting, it only persists based on if the form was posted and should be handled with local state.

**Comment by dbtlr** [comment]
With Cache Stats, are you talking about reformating this response:

GET /wp-json/wp-super-cache/v1/stats
```
{
  "supercache": {
    "expired": 0,
    "cached": 0,
    "fsize": 0
  },
    "wpcache": {
    "expired": 0,
    "cached": 0,
    "fsize": 0
  }
}
```

If so, I'm not sure what the URI property is supposed to return. Could you propose what the full response is supposed to look like here? We can definitely adjust this.

**Comment by dbtlr** [comment]
Regarding `is_super_cache_enabled` being set properly, this comes from the `$super_cache_enabled` global directly. This one definitely needs some input from @donnchawp, since he won't need to dig as much to find and answer.

**Comment by dbtlr** [comment]
In terms of the saving, I've refactored that endpoint a bit to be more explicit in the mapping of fields to saving methods and in the process uncovered some typos in these save field for `is_cache_enabled`.

The `cache_mod_rewrite` setting appears tied explicitly to the `is_super_cache_enabled`. I'm not sure why, maybe @donnchawp can comment on that as well. :)

**Comment by dbtlr** [comment]
@donnapep ^^^

**Comment by donnapep** [comment]
> With Cache Stats, are you talking about reformating this response:
> 
> GET /wp-json/wp-super-cache/v1/stats

Yes, I believe that is the endpoint for regenerating the cache stats, and it should return the updated cache stats with the structure defined below. The plan was to also return this data from the settings endpoint as well though to avoid having to make separate requests.

The format is as per the below:

```
cache_stats: {
  generated: 60,
  supercache: {
    cached: 1,
    cached_list: [
      {
        age: 3029,
        uri: 'localhost/blogs/',
      },
    ],
    expired: 1,
    expired_list: [
      {
        age: 4975,
        uri: 'localhost/support/',
      },
    ],
    fsize: '69.78KB',
  },
  wpcache: {
    cached: 1,
    cached_list: [
      {
        age: 95,
        uri: 'localhost/support-info/',
      },
    ],
    expired: 1,
    expired_list: [
      {
        age: 81,
        uri: 'localhost/blogs/',
      },
    ],
    fsize: '32.18KB',
  },
}
```

> If so, I'm not sure what the URI property is supposed to return.

URI would be the URI from the following screen capture of cached / expired files on the Contents tab:

![cached-files-uri](https://cloud.githubusercontent.com/assets/1190420/25385022/86415996-2990-11e7-805b-44a87ae19548.jpg)

**Comment by donnapep** [comment]
> In terms of the saving, I've refactored that endpoint a bit to be more explicit in the mapping of fields to saving methods and in the process uncovered some typos in these save field for is_cache_enabled.

To clarify, none of the settings were saving. `is_cache_enabled` was just an example. I did a bit of debugging on the plugin side and [this line](https://github.com/Automattic/wp-super-cache/blob/rest_api_settings_page/rest/class.wp-super-cache-rest-update-settings.php#L15) results in an empty array. I didn't debug any further than that, so I'm not sure if the problem is in the plugin, in Calypso, or both.

To test saving from Calypso, you can use [this branch](https://github.com/Automattic/wp-calypso/pull/13302).

Let me know if you need more info. Thx!

**Comment by donnapep** [comment]
`cache_gzencode` returns `null`. Does that sound correct? I'd thought it would return either `true` or `false`.

**Comment by dbtlr** [comment]
Definitely a work in progress. The internals of WPSC don't lend themselves very well to this type of work, so it is going to be a lot of trial and error to get there. I'll go through all of these today with your Calypso branch and see what I can do to fix as many of these as possible. :)

**Comment by donnchawp** [comment]
I added compression warnings. The notice will appear if either of these are true (or both true):

1. if WPSC_DISABLE_COMPRESSION is defined.
2. if the function gzencode() isn't defined.

If the notice is given you can hide the compression setting.

**Comment by donnchawp** [comment]
I need to remove cache_gzencode. Use cache_compression instead to enable or disable it.

**Comment by donnchawp** [comment]
cache_rejected_user_agent is what you should use for the rejected user agents.
cache_lock_down for lock_down status. You can set that now as well.

I fixed the "boolification" of those array settings (cache_acceptable_files and cache_rejected_uri)

**Comment by donnchawp** [comment]
You can use cache_compression to tell if gzip encoding is enabled or not.

**Comment by donnchawp** [comment]
In 48c7f0900539ea5058d4895be98f3c56515869ca I fixed saving of configuration settings. Update the settings on the preload page and `preload_on` and `preload_taxonomies` will be returned as bools.

In 9f1072e24fb70d7e57027592ce332266be3365b4I fixed saving of wp_cache_refresh_single_only (or refresh_current_only_on_comments)
In ca61ae4f40feea849f4a8b470c58104c426e63ec I added code to convert ossdl_https and refresh_current_only_on_comments to bools. refresh_current_only_on_comments has to be fixed this way because the error saving it as a string was in master for a long time so existing configurations will have that setting as a string. 
ossdl_https is saved as an option which returns a string.

**Comment by donnchawp** [comment]
The cache file listing looks similar to that above now (163bcfe52b9136793362c445cae8778cba17c437) but I changed it slightly so it shows directories instead of files, and lists the lower and upper age of file(s) in the directory. Deleting a directory (from wp-admin) deletes the supercache AND wpcache files in that directory.

**Comment by donnchawp** [comment]
As of bc9bba0c4600e47650b0586e584d40addd570756 you can set cache_compression to enable or disable cache compression/gzip encoding.

**Comment by donnchawp** [comment]
In 9ba185a705c16fe5abc67961c813cf9cf645f9e6 I fixed the problem with is_super_cache_enabled.

The plugin was setting it to false because of GET parameters in the URL. I fixed it by loading the configuration page in the "get settings" function. Had to take care of WPLOCKDOWN too, as it's a constant and you can't load constants twice without triggering a warning or error.

**Comment by donnchawp** [comment]
In 08917cd921cb331235ec1ce80a4da9f7888c9756 I added a "reset" API parameter that will reset the configuration, copying the sample configuration file over, and setting up some default garbage collection variables.

**Comment by donnchawp** [comment]
I removed the cache_stats setting and added the cache_type setting.
